src/dc_tools.rs

```diff
@@ -744,39 +744,34 @@ pub unsafe fn dc_create_id() -> *mut libc::c_char {
     encode_66bits_as_base64(buf[0usize], buf[1usize], buf[2usize])
 }
 
-/* ******************************************************************************
- * generate Message-IDs
- ******************************************************************************/
-unsafe fn encode_66bits_as_base64(v1: uint32_t, v2: uint32_t, fill: uint32_t) -> *mut libc::c_char {
-    /* encode 66 bits as a base64 string. This is useful for ID generating with short strings as
-    we save 5 character in each id compared to 64 bit hex encoding, for a typical group ID, these are 10 characters (grpid+msgid):
-    hex:    64 bit, 4 bits/character, length = 64/4 = 16 characters
-    base64: 64 bit, 6 bits/character, length = 64/6 = 11 characters (plus 2 additional bits) */
-    let ret: *mut libc::c_char = malloc(12) as *mut libc::c_char;
-    assert!(!ret.is_null());
-
-    static mut CHARS: [libc::c_char; 65] = [
+/// Encode 66 bits as a base64 string.
+/// This is useful for ID generating with short strings as we save 5 character
+/// in each id compared to 64 bit hex encoding. For a typical group ID, these
+/// are 10 characters (grpid+msgid):
+///    hex:    64 bit, 4 bits/character, length = 64/4 = 16 characters
+///    base64: 64 bit, 6 bits/character, length = 64/6 = 11 characters (plus 2 additional bits)
+/// Only the lower 2 bits of `fill` are used.
+fn encode_66bits_as_base64(v1: u32, v2: u32, fill: u32) -> String {
```

Comment:
if we make this safe how about just using the base64 crate instead. also we can change id to generate a bytebuffer instead and then just call base64::encode on it

Reply:
I will try it given that there is https://docs.rs/base64/0.10.1/base64/write/struct.EncoderWriter.html

Reply:
much nicer, thanks